Fixes a bug in WriteBuffer#shiftBytesLeft

zslayton opened the pull request:
This patch fixes a bug in the `WriteBuffer` that caused the
`shiftBytesLeft` method to corrupt data if it was called such that the
final `Block` in the list was emptied AND the second-to-last block
reclaimed some capacity. This would only occur if the writer
was configured to use length preallocation greater than 0 and if they
were writing more than 32KB of Ion (the default block size) between
flushes.

This change accounts for that case and, if the last `Block`
is emptied, returns it to the pool for reuse.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

almann commented on test/com/amazon/ion/impl/bin/WriteBufferTest.java:
```diff
@@ -950,6 +950,58 @@ public void shiftBytesLeftAcrossBufferBlocks() throws IOException {
         assertBuffer("0123489ABCDEF".getBytes());
     }
 
+    @Test
+    public void shiftBytesLeftAcrossBufferBlocksEmptyingLastBlock() throws IOException {
+        assertEquals(11, ALLOCATOR.getBlockSize());
+        // The "B" is the first and only byte in the second block.
+        // Shifting 1 byte left by one empties the last block.
+        buf.writeBytes("0123456789AB".getBytes());
+        buf.shiftBytesLeft(1, 1);
+        assertBuffer("0123456789B".getBytes());
+    }
```
Do/should we have tests for shifting bytes `==` to block size and `>=` block size?

zslayton replied:
I can add tests for those. In practice, the largest shifts are by `2` and the block size is 32KB. However, at the unit test level there's no reason I can't/shouldn't test that.

almann reviewed (commented):
Overall the change looks good--but is complex enough that I think we should consider more tests around various offsets and lengths around buffer boundaries.

tgregg commented:
I think the following test should pass, but it fails with `ArrayIndexOutOfBoundsException` on the line `buf.shiftBytesLeft(1, 6)`, for any shift amount >= 6 (which is just across the block boundary). It works for shift amounts of 5 or less (remaining in the same block).

```java
    @Test
    public void shiftBytesLeftAcrossBufferBlocksToStartThenWritePastBlockBoundary() throws IOException {
        assertEquals(11, ALLOCATOR.getBlockSize());
        // The "B" is the first and only byte in the second block.
        // Shifting 2 bytes left by 10 empties the last block and positions the shifted bytes at the start of the first
        // block.
        buf.writeBytes("0123456789AB".getBytes());
        buf.shiftBytesLeft(2, 10);
        assertBuffer("AB".getBytes());
        // Write enough bytes to expand back into the next block.
        buf.writeBytes("CDE0123456789AB".getBytes());
        assertBuffer("ABCDE0123456789AB".getBytes());
        // Shift one byte left past the block boundary. This should once again empty the last block.
        buf.shiftBytesLeft(1, 6);
        assertBuffer("ABCDE01234B".getBytes());
    }
```

@zslayton can you double check my test logic?

zslayton commented:
> 



> I think the following test should pass, but it fails with `ArrayIndexOutOfBoundsException` on the line `buf.shiftBytesLeft(1, 6)`, for any shift amount >= 6 (which is just across the block boundary). It works for shift amounts of 5 or less (remaining in the same block).
> 
> ```java
>     @Test
>     public void shiftBytesLeftAcrossBufferBlocksToStartThenWritePastBlockBoundary() throws IOException {
>         assertEquals(11, ALLOCATOR.getBlockSize());
>         // The "B" is the first and only byte in the second block.
>         // Shifting 2 bytes left by 10 empties the last block and positions the shifted bytes at the start of the first
>         // block.
>         buf.writeBytes("0123456789AB".getBytes());
>         buf.shiftBytesLeft(2, 10);
>         assertBuffer("AB".getBytes());
>         // Write enough bytes to expand back into the next block.
>         buf.writeBytes("CDE0123456789AB".getBytes());
>         assertBuffer("ABCDE0123456789AB".getBytes());
>         // Shift one byte left past the block boundary. This should once again empty the last block.
>         buf.shiftBytesLeft(1, 6);
>         assertBuffer("ABCDE01234B".getBytes());
>     }
> ```
> 
> @zslayton can you double check my test logic?

`shiftBytesLeft` is intended to shift the last `N` bytes to the left by up to `N` positions. (In practice, it's only called when `length >= 14` and `shiftBy <= 2`.) It's not clear that shifting `N` bytes by more than `N` positions should be a supported use case.

If you're alright with it, I think I'll address this by checking for this in `shiftBytesLeft` and throwing an `IllegalArgumentException` if it arises. I'll also note it in the docs and add a unit test for it.

tgregg commented:
> If you're alright with it, I think I'll address this by checking for this in shiftBytesLeft and throwing an IllegalArgumentException if it arises. I'll also note it in the docs and add a unit test for it.

Yep, that works for me.

zslayton commented:
I rewrote `shiftBytesLeftAcrossBlocks` from scratch and I think the updated version is _much_ easier to follow. The new version has far fewer branches and eliminates a lot of the restrictions that were in the original version. I've also added several new unit tests, including the one Tyler provided.

This last commit obviates a lot of the changes in the first commit. Instead of viewing "Changes since my last review", I suggest looking at all the commits together; it's more straightforward.

tgregg commented on test/com/amazon/ion/impl/bin/WriteBufferTest.java:
```diff
@@ -959,4 +959,118 @@ public void shiftBytesLeftAcrossBufferBlocksExclusively() throws IOException {
+        buf.writeBytes("0123456789|0123456789|0123456789|0123456789|0123456789|".getBytes());
+        // We can shift left amounts greater than the block size
+        buf.shiftBytesLeft(24, 24);
+        buf.writeBytes("0123456789|0123456789|012345678".getBytes());
```
I spent quite a while step-through debugging this test, because this assertion didn't look right. But the code was working perfectly...

Then I realized there's a reason this assertion wasn't failing... because it's not an assertion! The following assertion works:

```
assertBuffer("01234569|0123456789|0123456789|".getBytes());
```

zslayton replied:
🤦 I really should've slept on this commit. Don't code drowsy, kids.

tgregg commented on test/com/amazon/ion/impl/bin/WriteBufferTest.java:
```diff
@@ -959,4 +959,118 @@ public void shiftBytesLeftAcrossBufferBlocksExclusively() throws IOException {
+        buf.writeBytes("0123456789AB".getBytes());
+        // Shift left by more bytes than we're shifting (shiftBy > length)
+        buf.shiftBytesLeft(2, 5);
+        buf.writeBytes("01234AB".getBytes());
```
Replace `buf.writeBytes` with `assertBuffer` here too.

tgregg commented on test/com/amazon/ion/impl/bin/WriteBufferTest.java:
```diff
@@ -959,4 +959,118 @@ public void shiftBytesLeftAcrossBufferBlocksExclusively() throws IOException {
+        assertEquals(11, ALLOCATOR.getBlockSize());
+        buf.writeBytes("01234567AB".getBytes());
+        buf.shiftBytesLeft(2, 8);
+        buf.writeBytes("AB".getBytes());
```
Replace `buf.writeBytes` with `assertBuffer` here too.

tgregg commented on test/com/amazon/ion/impl/bin/WriteBufferTest.java:
```diff
@@ -959,4 +959,118 @@ public void shiftBytesLeftAcrossBufferBlocksExclusively() throws IOException {
+        assertEquals(11, ALLOCATOR.getBlockSize());
+        buf.writeBytes("0123456789AB".getBytes());
+        buf.shiftBytesLeft(2, 10);
+        buf.writeBytes("AB".getBytes());
```
Replace `buf.writeBytes` with `assertBuffer` here too.

jobarr-amzn commented on test/com/amazon/ion/impl/bin/WriteBufferTest.java:
```diff
@@ -959,4 +959,118 @@ public void shiftBytesLeftAcrossBufferBlocksExclusively() throws IOException {
+    public void shiftBytesLeftByMoreThanTheBlockSize() {
+        assertEquals(11, ALLOCATOR.getBlockSize());
+        // We have 5 blocks' worth of data
+        buf.writeBytes("0123456789|0123456789|0123456789|0123456789|0123456789|".getBytes());
+        // We can shift left amounts greater than the block size
+        buf.shiftBytesLeft(24, 24);
+        assertBuffer("01234569|0123456789|0123456789|".getBytes());
```
Borrowing a convention from counting time in music, I think this makes the transpositions easier to see. 
```suggestion
        buf.writeBytes("0123456789|1123456789|2123456789|3123456789|4123456789|".getBytes());
        // We can shift left amounts greater than the block size
        buf.shiftBytesLeft(24, 24);
        assertBuffer("01234569|3123456789|4123456789|".getBytes());
```

zslayton commented on src/com/amazon/ion/impl/bin/WriteBuffer.java:
```diff
@@ -227,63 +227,72 @@ private void shiftBytesLeftWithinASingleBlock(int length, int shiftBy) {
+        // This is the buffer offset of the first byte that we will be shifting backwards.
+        long sourceBufferOffset = position() - length;
+        // When we're done, this will be the first offset in the buffer that does not contain data.
+        long writeBufferLimit = sourceBufferOffset - shiftBy + length;
```
> And isn't this the same as position() - shiftBy? Would that be more clear?

I like that better, thanks.
